### PR TITLE
Add a new dialog-driven way to generate a constructor.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @Pilchie @dpoeschl

**Comment by CyrusNajmabadi** [comment]
Experience looks like this:

repro

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 1]
```diff
@@ -1178,4 +1178,10 @@ This version used in: {2}</value>
   <data name="Remove_separators" xml:space="preserve">
     <value>Remove separators</value>
   </data>
-</root>
+  <data name="Generate_constructor" xml:space="preserve">
+    <value>Generate constructor...</value>
+  </data>
+  <data name="Pick_members_to_be_used_as_constructor_parameters" xml:space="preserve">
+    <value>Pick members to be used as constructor parameters</value>
```
Comment (dpoeschl) [inline src/Features/Core/Portable/FeaturesResources.resx]: "Choose and order members to be used..."?

File: src/Features/Core/Portable/GenerateConstructorFromMembers/GenerateConstructorFromMembersCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -31,6 +36,52 @@ internal partial class GenerateConstructorFromMembersCodeRefactoringProvider : A
...
+                    var containingType = GetEnclosingNamedType(semanticModel, root, textSpan.Start, cancellationToken);
+                    if (containingType?.TypeKind == TypeKind.Class || containingType?.TypeKind == TypeKind.Struct)
+                    {
+                        if (!containingType.IsStatic)
+                        {
+                            var viableMembers = containingType.GetMembers().WhereAsArray(IsWritableInstanceFieldOrProperty);
+                            if (viableMembers.Length == 0 &&
+                                containingType.InstanceConstructors.Any(c => c.Parameters.Length == 0 && !c.IsImplicitlyDeclared))
+                            {
+                                // If there are no fields, and there's already an explicit empty parameter constructor
```
Comment (dpoeschl) [inline src/Features/Core/Portable/GenerateConstructorFromMembers/GenerateConstructorFromMembersCodeRefactoringProvider.cs]: "default" or "parameterless"

File: src/Features/Core/Portable/GenerateConstructorFromMembers/GenerateConstructorFromMembersCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -42,22 +93,99 @@ internal partial class GenerateConstructorFromMembersCodeRefactoringProvider : A
...
+                    result.Members, cancellationToken);
+
+                // There was an existing constructor that matched what the user wants to create.
+                // Generate it if it's the implicit, no-arg, constructor, otherwise just navigate
+                // to the existing constructor
+                if (state.MatchingConstructor != null)
+                {
+                    if (state.MatchingConstructor.IsImplicitlyDeclared)
+                    {
+                        var codeAction = new FieldDelegatingCodeAction(_service, _document, state);
```
Comment (dpoeschl) [inline src/Features/Core/Portable/GenerateConstructorFromMembers/GenerateConstructorFromMembersCodeRefactoringProvider.cs]: Can we call this `FieldAssigningCodeAction` or something? Nothing is being delegated through a field or anything like that.

File: src/Features/Core/Portable/GenerateFromMembers/AbstractGenerateFromMembersService.cs [hunk 4]
```diff
@@ -132,19 +116,11 @@ where constructorTypes.SequenceEqual(symbolTypes)
...
         }
 
-        protected bool HasMatchingConstructor(
-            INamedTypeSymbol containingType,
-            List<IParameterSymbol> parameters)
-        {
-            return containingType.InstanceConstructors.Any(c => MatchesConstructor(c, parameters));
-        }
+        protected IMethodSymbol GetMatchingConstructor(INamedTypeSymbol containingType, ImmutableArray<IParameterSymbol> parameters)
+            => containingType.InstanceConstructors.FirstOrDefault(c => MatchesConstructor(c, parameters));
```
Comment (dpoeschl) [inline src/Features/Core/Portable/GenerateFromMembers/AbstractGenerateFromMembersService.cs]: Maybe explain that the FirstOrDefault is to pick between identical constructors in error code.

File: src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialogViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+        private void Move(List<MemberSymbolViewModel> list, int index, int delta)
+        {
+            var param = list[index];
+            list.RemoveAt(index);
+            list.Insert(index + delta, param);
+
+            SelectedIndex += delta;
+
+            //NotifyPropertyChanged(nameof(AllParameters));
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialogViewModel.cs]: Remove commented out code.

**Comment by CyrusNajmabadi** [comment]
I would like to do integration tests in a followup PR.  Can you help me write against this new system we have (Since you have experience with this with CHangeSig)?

**Comment by Pilchie** [comment]
@CyrusNajmabadi Why? All that does is increase test debt, and the odds that you end up getting distracted by something else and never coming back to it.

**Comment by CyrusNajmabadi** [comment]
Alright, i'll work with david tomorrow to see how to properly integration test this guy.

**Comment by CyrusNajmabadi** [comment]
Unit tests for C# added.

**Comment by CyrusNajmabadi** [comment]
@dpoeschl VB side is done, as are unit tests.  Can you help me tomorrow with the integration test side of things?

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest please

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs [hunk 6]
```diff
@@ -627,5 +629,151 @@ public Contribution(string title, int number)
...
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateConstructorFromMembers)]
+        public async Task TestSingleFieldWithDialog()
+        {
+            await TestWithGenerateConstructorDialogAsync(
+@"using System.Collections.Generic;
+
+class Z
+{
+    int a;
+    [||]
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs]: So, we're going to offer this everywhere inside a class?

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs [hunk 7]
```diff
@@ -627,5 +629,151 @@ public Contribution(string title, int number)
...
+            await TestMissingAsync(
+@"using System.Collections.Generic;
+
+class Z
+{
+    int a;
+    string b;
+    public void M()
+    {
+    }[||]
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs]: What about inside members?

File: src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs [hunk 8]
```diff
@@ -10,33 +11,39 @@
...
 using Xunit;
-using Microsoft.CodeAnalysis.UnitTests;
 
 namespace Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions
 {
     public abstract class AbstractCodeActionTest : AbstractCodeActionOrUserDiagnosticTest
     {
-        protected abstract CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace);
+        protected virtual CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace)
+            => throw new NotImplementedException();
```
Comment (Pilchie) [inline src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs]: Why not just remove?  Looks like you updated callers anyway.

File: src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs [hunk 9]
```diff
@@ -105,5 +112,24 @@ protected static Document GetDocument(TestWorkspace workspace)
...
+            private readonly ImmutableArray<string> _memberNames;
+
+            public TestPickMembersService(ImmutableArray<string> memberNames)
+                => _memberNames = memberNames;
+
+            public PickMembersResult PickMembers(string title, ImmutableArray<ISymbol> members)
+                => new PickMembersResult(_memberNames.SelectAsArray(n => members.Single(m => m.Name == n)));
+        }
+
+        protected Task TestWithGenerateConstructorDialogAsync(
```
Comment (Pilchie) [inline src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs]: This seems like the wrong file to contain this method.

File: src/Features/Core/Portable/GenerateConstructorFromMembers/ConstructorDelegatingCodeAction.cs [hunk 10]
```diff
@@ -71,9 +73,10 @@ private class ConstructorDelegatingCodeAction : CodeAction
                         parameters: _state.Parameters,
                         statements: statements,
                         thisConstructorArguments: thisConstructorArguments),
-                    new CodeGenerationOptions(contextLocation: syntaxTree.GetLocation(_state.TextSpan)),
-                    cancellationToken: cancellationToken)
-                    .ConfigureAwait(false);
+                    new CodeGenerationOptions(
+                        contextLocation: syntaxTree.GetLocation(_state.TextSpan),
+                        afterThisLocation: _state.TextSpan.IsEmpty ? syntaxTree.GetLocation(_state.TextSpan) : null),
```
Comment (Pilchie) [inline src/Features/Core/Portable/GenerateConstructorFromMembers/ConstructorDelegatingCodeAction.cs]: This looks backwards to me?

Comment (Pilchie) [inline src/Features/Core/Portable/GenerateConstructorFromMembers/ConstructorDelegatingCodeAction.cs]: Or else, I don't know what this is trying to do.

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs]: Nope.  Only on completely blank lines between members.  Which i think is reasonable. :)

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs (hunk 7 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs]: Nope.  not inside members, only on completely whitespace lines between members.

Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs]: My point was that there are no tests which verify that.

File: src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs]: this approach allowed all the existing overrides to not have to be updated... but i could just go and update them all...

File: src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs (hunk 7 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateConstructorFromMembers/GenerateConstructorFromMembersTests.cs]: ah, i see.  i felt that this was sufficient.  But i could add another test.

File: src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs]: This gets refactored in the GenerateEqualsAndGetHashCode tests in the next PR.  The idea is that this is the helper function that allows us to test the feature while swapping in a mock IPickMembersService.

File: src/Features/Core/Portable/GenerateConstructorFromMembers/ConstructorDelegatingCodeAction.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateConstructorFromMembers/ConstructorDelegatingCodeAction.cs]: yeah, i should def comment ths :)  

General idea: if you're selecting a bunch of members, and generating a constructor, then we'll auto find the right place for the constructor.

However, if you're just invoking the feature at a particular location, then we'll insert at that location point.

File: src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialogViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+            {
+                if (!CanMoveDown)
+                {
+                    return string.Empty;
+                }
+
+                return string.Format(ServicesVSResources.Move_0_below_1, MemberContainers[SelectedIndex.Value].MemberAutomationText, MemberContainers[SelectedIndex.Value + 1].MemberAutomationText);
+            }
+        }
+
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialogViewModel.cs]: Extra blank lines.

File: src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialogViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        public List<MemberSymbolViewModel> MemberContainers { get; set; }
+
+        internal class MemberSymbolViewModel : AbstractNotifyPropertyChanged
+        {
+            private readonly IGlyphService _glyphService;
+
+            public ISymbol MemberSymbol { get; }
+
+            private static SymbolDisplayFormat s_memberDisplayFormat = new SymbolDisplayFormat(
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialogViewModel.cs]: Aren't these fields/properties?  Do this items make sense?

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicGenerateConstructorDialog.cs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.IntegrationTest.Utilities;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Input;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Roslyn.VisualStudio.IntegrationTests.VisualBasic
+{
+    [Collection(nameof(SharedIntegrationHostFixture))]
+    public class BasicGenerateConstructorDialog : AbstractEditorTest
```
Comment (Pilchie) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicGenerateConstructorDialog.cs]: No C# integration test?

File: src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialog.xaml [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+             xmlns:u="clr-namespace:Microsoft.VisualStudio.LanguageServices.Implementation.Utilities"
+             xmlns:imaging="clr-namespace:Microsoft.VisualStudio.Imaging;assembly=Microsoft.VisualStudio.Imaging"
+             xmlns:imagecatalog="clr-namespace:Microsoft.VisualStudio.Imaging;assembly=Microsoft.VisualStudio.ImageCatalog"
+             mc:Ignorable="d" 
+             d:DesignHeight="380" d:DesignWidth="460"
+             Height="380" Width="460"
+             MinHeight="380" MinWidth="460"
+             Title="{Binding ElementName=dialog, Path=PickMembersDialogTitle}"
+             HasHelpButton="True"
+             FocusManager.FocusedElement="{Binding ElementName=interfaceNameTextBox}"
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/PickMembers/PickMembersDialog.xaml]: Copy/paste!

**Pull request merged**